taca_ngi_pipeline/deliver/deliver_castor.py

```diff
@@ -0,0 +1,170 @@
+                self.update_delivery_status(status="DELIVERED")
+                self.acknowledge_delivery()
+            return status
+        except (db.DatabaseError, DelivererInterruptedError, Exception):
+            raise
+
+
+
+
+
+class CastorSampleDeliverer(SampleDeliverer):
+    """
+        A class for handling sample deliveries to castor
+    """
+
+    def __init__(self, projectid=None, sampleid=None, sftp_client=None, **kwargs):
+        super(CastorSampleDeliverer, self).__init__(
+            projectid,
+            sampleid,
+            **kwargs)
+        self.sftp_client = sftp_client
+
+
+    def do_delivery(self):
+        """ Deliver the staged delivery folder using sftp
+            :returns: True if delivery was successful, False if unsuccessful
+            :raises DelivererTOBEDEFINEDError: if an exception occurred during
+                transfer
+        """
+        # transfer it (maybe an open session is needed)
+        logger.info("{} transferring sample to castor sftp server".format(self.sampleid))
+        try:
+            #http://stackoverflow.com/questions/4409502/directory-transfers-on-paramiko
+            #walk through the staging path and recreate the same path in castor and put files there
+            origin_folder_sample = os.path.join(self.expand_path(self.stagingpath), self.sampleid)
+            #create the sample folder
+            if not rexists(self.sftp_client, self.sampleid):
+                self.sftp_client.mkdir(self.sampleid)
+            #now target dir is created
+            targed_dir = self.sampleid
+            self.sftp_client.put_dir(origin_folder_sample ,targed_dir)
+        except Exception as e:
+            print 'Caught exception: {}: {}'.format(e.__class__, e)
+            raise
+        logger.info("{} sample transferred to castor sftp server".format(self.sampleid))
+        # return True, if something went wrong an exception is thrown before this
+        return True
+
+###http://stackoverflow.com/questions/4409502/directory-transfers-on-paramiko
+class MySFTPClient(paramiko.SFTPClient):
+    def put_dir(self, source, target):
+        ''' Uploads the contents of the source directory to the target path. The
+            target directory needs to exists. All subdirectories in source are 
+            created under target.
+        '''
+        for item in os.listdir(source):
+            if os.path.isfile(os.path.join(source, item)):
+                self.put(os.path.join(source, item), '%s/%s' % (target, item))
+            else:
+                self.mkdir('%s/%s' % (target, item), ignore_existing=True)
+                self.put_dir(os.path.join(source, item), '%s/%s' % (target, item))
+
+    def mkdir(self, path, mode=511, ignore_existing=False):
+        ''' Augments mkdir by adding an option to not fail if the folder exists  '''
+        try:
+            super(MySFTPClient, self).mkdir(path, mode)
```

Comment:
The try close is needed to deal with the case folder already exists. This class implements the recursive copy of a folder by iterating over the `source` directory